atlasdb-cassandra/src/test/java/com/palantir/atlasdb/cassandra/ReloadingCloseableContainerTest.java

```diff
@@ -0,0 +1,328 @@
+        container.close();
+    }
+
+    @Test
+    public void blockingMethodsDontDeadlockUnderConcurrentLoad() {
+        int numberOfRuns = 2000;
+
+        ReloadingCloseableContainer<Closeable> container = getContainer();
+
+        AtomicInteger refreshableValue = new AtomicInteger(INITIAL_VALUE);
+        CompletableFuture<?> allFutures =
+                runManyTimesAndGetSingleFuture(numberOfRuns, () -> triggerEvent(container, refreshableValue));
+
+        Awaitility.await().atMost(Duration.ofSeconds(2)).until(allFutures::isDone);
+
+        Future<?> future = executor.submit(container::close);
+        Awaitility.await().atMost(Duration.ofSeconds(1)).until(future::isDone);
+    }
+
+    @Test
+    public void runWithResourceReturnsFunctionValue() {
+        ReloadingCloseableContainer<Closeable> container = getContainer();
+        int expectedValue = 3213;
+        int returnedValue = container.runWithResource(_closeable -> expectedValue);
+        assertThat(returnedValue).isEqualTo(expectedValue);
+    }
+
+    @Test
+    public void runWithResourceRunsWithLatestResource() {
+        ReloadingCloseableContainer<Closeable> container =
+                ReloadingCloseableContainer.of(closeableFactory, settableRefreshable);
+        settableRefreshable.update(UPDATED_VALUE);
+        container.executeWithResource(closeable -> assertThat(closeable).isEqualTo(mockCloseableExtra));
+    }
+
+    @Test
+    public void runWithResourcePropagatesException() {
+        ReloadingCloseableContainer<Closeable> container = getContainer();
+        assertThatThrownBy(() -> container.runWithResource(_closeable -> {
+                    throw new SafeIllegalStateException("Test");
+                }))
+                .isInstanceOf(SafeIllegalStateException.class);
+    }
+
+    @Test
+    public void runWithDoesNotBlockResourceUpdates() {
+        ReloadingCloseableContainer<Closeable> container = getContainer();
+        CyclicBarrier ensureInitialExecuteHasStartedBarrier = new CyclicBarrier(2);
+        CyclicBarrier manuallyFinishRunWithBarrier = new CyclicBarrier(2);
+        executor.submit(() -> container.executeWithResource(closeable -> {
+            awaitOrFail(ensureInitialExecuteHasStartedBarrier);
+            awaitOrFail(manuallyFinishRunWithBarrier);
+            assertThat(closeable).isEqualTo(mockCloseable);
+        }));
+
+        awaitOrFail(ensureInitialExecuteHasStartedBarrier);
+        // Map and subscriber updates happen sequentially, but on the same thread. Therefore, the next call won't
+        // complete until after the execute above completes, even if the map itself has finished.
+        executor.submit(() -> settableRefreshable.update(UPDATED_VALUE));
+
+        // The earlier executeWithResource call cannot complete yet, and so if the update was blocked, this following
+        // assertion would fail.
+
+        // It is possible for this to fail, if this call ends up scheduled before the update has started...
```

Comment:
Ditto on the other utilities comment - will see if I can make the order deterministic